ibis/expr/api.py

```diff
@@ -1947,6 +1949,102 @@ def watermark(time_col: str, allowed_delay: ir.IntervalScalar) -> Watermark:
     return Watermark(time_col=time_col, allowed_delay=allowed_delay)
 
 
+@functools.singledispatch
+def range(start, stop, step) -> ir.ArrayValue:
+    """Generate a range of values.
+
+    ::: {.callout-note}
+    `start` is inclucive and `stop` is exclusive, just like Python's builtin
+    [`range`](range).
+
+    When `step` equals 0, however, this function will return an empty array.
+
+    Python's `range` will raise an exception when `step` is zero.
+    :::
+
+    Parameters
+    ----------
+    start
+        Lower bound of the range, inclusive.
+    stop
+        Upper bound of the range, exclusive.
+    step
+        Step value. Optional, defaults to 1.
+
+    Returns
+    -------
+    ArrayValue
+        An array of values
+
+    Examples
+    --------
+    >>> import ibis
+    >>> ibis.options.interactive = True
+
+    Range using only a stop argument
+
+    >>> ibis.range(5)
+    [0, 1, 2, 3, 4]
+
+    Simple range using start and stop
+
+    >>> ibis.range(1, 5)
+    [1, 2, 3, 4]
+
+    Generate an empty range
+
+    >>> ibis.range(0)
+    []
+
+    Negative step values are supported
+
+    >>> ibis.range(10, 4, -2)
+    [10, 8, 6]
+
+    `ibis.range` behaves the same as Python's range ...
+
+    >>> ibis.range(0, 7, -1)
+    []
+
+    ... except when the step is zero, in which case `ibis.range` returns an
+    empty array
+
+    >>> ibis.range(0, 5, 0)
+    []
+
+    Because the resulting expression is array, you can unnest the values
+
+    >>> ibis.range(5).unnest().name("numbers")
+    ┏━━━━━━━━━┓
+    ┃ numbers ┃
+    ┡━━━━━━━━━┩
+    │ int8    │
+    ├─────────┤
+    │       0 │
+    │       1 │
+    │       2 │
+    │       3 │
+    │       4 │
+    └─────────┘
+    """
+    raise NotImplementedError()
+
+
+@range.register(int)
```

Comment:
Shouldn't this return with an array literal instead?

Reply:
Why? It's perfectly fine to run `ibis.range(start, stop)` where `start` and `stop` are arbitrary integer expressions.

Reply:
Ohh, this is just overloading the first argument. I was thinking of a simplification where all the arguments passed to `ibis.range()` are integers where the result is equivalent with `ibis.literal(range(...))`.

Reply:
Sure, but I will not be bending over backwards to make the type annotations more specific, especially since they are correct as is :)